README.md
# Empoloyee
员工管理系统


员工管理系统源码/history.txt
zhangsan,root_history,,0,,0,,0,0,0,2017-6-21  2:5:8
lisi,user_update,,0,,0,,0,0,0,2017-6-21  2:20:7
lisi,user_update,,0,,0,,0,0,0,2017-6-21  2:21:40
zhangsan,root_insert,,0,,0,,0,0,0,2017-6-21  2:26:42
zhangsan,root_insert,,0,,0,,0,0,0,2017-6-21  2:30:41
lisi,user_update,lisi,1,1,1,11,1,1,1,2017-6-21  2:32:10
jinzhenyu,user_update,,0,,0,,0,0,0,2017-6-21  6:45:27
zhangsan,root_find,,0,,0,,0,0,0,2017-6-21  8:0:39
zhangsan,root_delete,,0,,0,,0,0,0,2017-6-21  8:1:50
zhangsan,root_find,,0,,0,,0,0,0,2017-6-21  8:2:5
jinzhenyu,user_sign_in,,0,,0,,0,0,0,2017-6-21  8:54:31
jinzhenyu,user_sign_in,,0,,0,,0,0,0,2017-6-21  8:56:18
jinzhenyu,user_sign_in,,0,,0,,0,0,0,2017-6-21  8:58:51
zhangsan,root_history,,0,,0,,0,0,0,2017-6-21  17:7:25
zhangsan,quit,,0,,0,,0,0,0,2017-6-21  17:26:8
zhangsan,quit,,0,,0,,0,0,0,2017-6-21  17:27:10
jinzhenyu,user_find,,0,,0,,0,0,0,2017-6-21  17:48:30
zhangsan,root_find,,0,,0,,0,0,0,2017-6-21  17:51:36
jinzhenyu,user_find,,0,,0,,0,0,0,2017-6-21  17:52:37
jinzhenyu,user_find,,0,,0,,0,0,0,2017-6-21  17:52:47
jinzhenyu,user_find,,0,,0,,0,0,0,2017-6-21  17:56:3
zhangsan,root_find,5,53,5,53,5,53,53,53,2017-6-21  20:38:59
zhangsan,root_find,5,53,5,53,5,53,53,53,2017-6-21  20:43:46


员工管理系统源码/server.c
#include "head.h"

struct worker_info_t worker;
struct login_info_t login;//从客户端接收的用户名和密码
struct login_info_t login_root = {//自定义的root用户的用户名和密码
    .name = "zhangsan",
    .password = "123"
};

int sockfd, acceptfd;
int len_login = sizeof(login);
int len_worker = sizeof(worker);
struct select select_t;
char buf[N] = {};
char op[N] = {};
sqlite3 *db;

/*登录验证函数
 *函数功能：检查来自客户端的用户名及密码是否与数据库中的匹配
 *          匹配之后确定其权限。
 *          login.flags = 1 root用户
 *          login.flags = 0 普通用户
 *          login.flags = -1 用户名或密码错误
 *函数参数：无
 *函数返回值：0 
 * */
int login_check(void)
{
    /*验证是否为root用户*/
    if((strcmp(login.name,login_root.name)==0) && (strcmp(login.password,login_root.password)==0))
    {
        login.flags = 1;
        if(send(acceptfd, (void *)&login, len_login, 0) <= 0)
        {
            perror("fail to send login_flags\n");
            exit(-1);
        }
        return 0;
    }

    /*验证是否为普通用户，用户名和密码需要从数据库中读取*/
    char *errmsg;
    char **result;
    int row, column;
    char sqlstr[128];
    sprintf(sqlstr,"select password from login where name='%s'",login.name);

    if(sqlite3_get_table(db, sqlstr, &result, &row, &column, &errmsg) != SQLITE_OK)
    {
        printf("%s\n", errmsg);
    }

    if(strcmp(login.password,result[1]) == 0)
    {
        login.flags = 0;
        if(send(acceptfd, (void *)&login, len_login, 0) <= 0)
        {
            perror("fail to send login_flags\n");
            exit(-1);
        }
    }

    /*用户名及密码与数据库不匹配，重新登录*/
    else 
    {
        login.flags = -1;
        if(send(acceptfd, (void *)&login, len_login, 0) <= 0)
        {
            perror("fail to send login_flags\n");
            exit(-1);
        }
    }
    
    return 0;
}

int do_insert(sqlite3 *db)
{

    if(recv(acceptfd, (void *)&worker, len_worker, 0) <= 0)
    {	
		perror("fail to receive worker information\n");
    	exit(-1);
	}

    if(recv(acceptfd, (void *)&login, len_login, 0) <= 0)
    {	
		perror("fail to receive login information\n");
    	exit(-1);
	}

    char sqlstr1[128],sqlstr2[128], *errmsg;
    bzero(buf,N);
    sprintf(sqlstr1, "insert into empoloyee values ('%s', '%s', %d, %d, '%s', %d, %d, %d)", worker.name, worker.sex, worker.age, worker.id, worker.section, worker.salary, worker.sign_in, worker.performance);
    sprintf(sqlstr2, "insert into login values ('%s', '%s')",login.name,login.password); 
    if (sqlite3_exec(db, sqlstr1, NULL, NULL, &errmsg) != 0)
    {
        strcpy(buf, "insert failed!");
        send(acceptfd, buf, N, 0);
        printf("error : %s\n", sqlite3_errmsg(db));
    }
    if (sqlite3_exec(db, sqlstr2, NULL, NULL, &errmsg) != 0)
    {
        strcpy(buf, "insert failed!");
        send(acceptfd, buf, N, 0);
        printf("error : %s\n", sqlite3_errmsg(db));
    }
    else
    {
        strcpy(buf, "insert ok!\n");
        send(acceptfd, buf, N, 0);
        printf("%s\n", buf);
    }
    return 0;
}
int do_delete(sqlite3 *db)
{
    bzero(buf,N);
    if(recv(acceptfd, buf, N, 0) <= 0)
    {	
		perror("fail to receive delete information\n");
    	exit(-1);
	}

    char *errmsg;
    char sqlstr[128];
    sprintf(sqlstr, "delete from empoloyee where name='%s'", buf);
    if (sqlite3_exec(db, sqlstr, NULL, NULL, &errmsg) != 0)
    {
        bzero(buf,N);
        strcpy(buf,"delete failed!\n");
        send(acceptfd,buf, N, 0);
        printf("error : %s\n", sqlite3_errmsg(db));
    }
    else 
    {
        bzero(buf,N);
        strcpy(buf,"delete ok!\n");
        send(acceptfd, buf, N, 0);
        printf("%s\n", buf);
    }

    return 0;
}

int do_root_update(sqlite3 *db)
{
	int num  = 0, new_num, new_score;
	if(recv(acceptfd, (void *)&select_t, sizeof(select_t), 0) <= 0)
	{
		perror("fail to receive\n");
		exit(-1);
	}
	char *errmsg;
	char sqlstr[N];
	char update_date[N];
	switch (select_t.num)
	{
	case 1 :
		sprintf(update_date,"id = %d", select_t.id);
		sprintf(sqlstr, "update empoloyee set %s where name = '%s'", update_date, select_t.o_name);
		break;
	case 2 :
		sprintf(update_date,"name = '%s'", select_t.n_name);
		sprintf(sqlstr, "update empoloyee set %s where name = '%s'", update_date, select_t.o_name);
		break;			
	case 3 :
		sprintf(update_date,"age = %d", select_t.age);
		sprintf(sqlstr, "update empoloyee set %s where name = '%s'", update_date, select_t.o_name);
		break;
	}
	if (sqlite3_exec(db, sqlstr, NULL, NULL, &errmsg) != 0)
	{
        bzero(buf, N);
        strcpy(buf, "root_update failed!");
        send(acceptfd, buf, N, 0);
		printf("error : %s\n", sqlite3_errmsg(db));
	}
	else 
    {
        bzero(buf, N);
        strcpy(buf, "root_update ok!");
        send(acceptfd, buf, N, 0);
		printf("root_update is done\n");
    }
	return 0;
    
}

int do_root_find(sqlite3 *db)
{
    #if 1
    bzero(buf, N);
	if(recv(acceptfd, buf, N, 0) <= 0)
	{
		perror("fail to receive root_find information!\n");
		exit(-1);
	}
    printf("%s\n",buf);
    #endif

    char *errmsg;
    char **result;
    char sqlstr[128] = {};
    int row, column, i, j, index;

    sprintf(sqlstr,"select * from empoloyee where name='%s'",buf);
    if(sqlite3_get_table(db, sqlstr, &result, &row, &column, &errmsg) != SQLITE_OK)
    {
        printf("%s\n", errmsg);
    }
    index = column;

    for(i = 0; i < column; i++)
    {
        printf("%-10s", result[i]);
    }
    putchar(10);
    for(j = 0; j < row; j++)
    {
        for(i = 0; i < column; i++)
        {
            printf("%-10s", result[index++]);
        }

        putchar(10);
    }
    strcpy(worker.name, result[column]);
    strcpy(worker.sex, result[column+1]);
    worker.age = *result[column+2];
    worker.id = *result[column+3];
    strcpy(worker.section, result[column+4]);
    worker.salary = *result[column+5];
    worker.sign_in = *result[column+6];
    worker.performance = *result[column+7];

	if(send(acceptfd, (void *)&worker, len_worker, 0) <= 0)
	{
		perror("fail to send root_find information!\n");
		exit(-1);
	}
	printf("root_find ok!\n");
    return 0;
}
/*
 *  发送日志
 */
int do_root_history()
{
	int fd, nbyte;
	char buf[32] = {0};
	if((fd = open("./history.txt", O_RDONLY)) < 0)
	{
		perror("fail to open history.txt");
		return -1;
	}
	while((nbyte = read(fd, buf, 31)) > 0)
	{
		send(acceptfd, buf, 31, 0);
	}
    bzero(buf, N);
	memset(buf,0,sizeof(buf));
	strcpy(buf,"end");
	send(acceptfd, buf, 31, 0);
}

int do_log(void)
{
	time_t tm;
	struct tm *mytm;
	time(&tm);  //获取秒数
	mytm = localtime(&tm);//将描述转化为年月日十分秒
	
	FILE *fd_history;


	fd_history = fopen("./history.txt","a+");
	if(fd_history == NULL)
	{
		perror("fail to open history.txt");
		return -1;
	}
	//打印操作用户及操作方式。
	fprintf(fd_history,"%s,%s,",login.name , op);
	fprintf(fd_history,"%s,%d,%s,%d,%s,%d,%d,%d,",worker.name, worker.age, worker.sex, worker.id, worker.section, worker.salary, worker.sign_in, worker.performance);
	//打印时间
	fprintf(fd_history,"%d-%d-%d  %d:%d:%d\n",mytm->tm_year+1900, mytm->tm_mon+1, mytm->tm_mday,
			mytm->tm_hour, mytm->tm_min, mytm->tm_sec);
	fflush(fd_history);	
	return 0;
}

int do_user_update(sqlite3 *db)
{
	int num  = 0, new_num, new_score;
	if(recv(acceptfd, (void *)&select_t, sizeof(select_t), 0) <= 0)
	{
		perror("fail to receive\n");
		exit(-1);
	}
	char *errmsg;
	char sqlstr[N];
	char update_date[N];
	switch (select_t.num)
	{
	case 1 :
		sprintf(update_date,"id = %d", select_t.id);
		sprintf(sqlstr, "update empoloyee set %s where name = '%s'", update_date, select_t.o_name);
		break;
	case 2 :
		sprintf(update_date,"password = '%s'", select_t.n_name);
		sprintf(sqlstr, "update login set %s where name = '%s'", update_date, select_t.o_name);
		break;			
	case 3 : 
		sprintf(update_date,"age = %d", select_t.age);
		sprintf(sqlstr, "update empoloyee set %s where name = '%s'", update_date, select_t.o_name);
		break;
	}
	if (sqlite3_exec(db, sqlstr, NULL, NULL, &errmsg) != 0)
	{
        bzero(buf, N);
        strcpy(buf, "root_update failed!");
        send(acceptfd, buf, N, 0);
		printf("error : %s\n", sqlite3_errmsg(db));
	}
	else 
    {
        bzero(buf, N);
        strcpy(buf, "root_update ok!");
        send(acceptfd, buf, N, 0);
		printf("root_update is done\n");
	}
    return 0;
}

int do_user_find(sqlite3 *db)
{
    char *errmsg;
    char **result;
    char sqlstr[128] = {};
    int row, column, i, index;

    sprintf(sqlstr,"select * from empoloyee where name='%s'",login.name);
    if(sqlite3_get_table(db, sqlstr, &result, &row, &column, &errmsg) != SQLITE_OK)
    {
        printf("%s\n", errmsg);
    }

    index = column;

    for(i = 0; i < column; i++)
    {
        printf("%-10s", result[i]);
    }
    putchar(10);
    for(i = 0; i < column; i++)
    {
        printf("%-10s", result[index++]);
    }            
    putchar(10);

    strcpy(worker.name, result[column]);
    strcpy(worker.sex, result[column+1]);
    worker.age = *result[column+2];
    worker.id = *result[column+3];
    strcpy(worker.section, result[column+4]);
    worker.salary = *result[column+5];
    worker.sign_in = *result[column+6];
    worker.performance = *result[column+7];
	if(send(acceptfd, (void *)&worker, len_worker, 0) <= 0)
	{
		perror("fail to send user_find information!\n");
		exit(-1);
	}
	printf("user_find ok!\n");
    return 0;

}

int do_user_sign_in(sqlite3 *db)
{
    if(recv(acceptfd, (void *)&worker, len_worker, 0) <= 0)
    {	
		perror("fail to receive worker information\n");
    	exit(-1);
	}

	char *errmsg;
	char sqlstr[128];

	sprintf(sqlstr, "update empoloyee set signin=%d where name = '%s'", worker.sign_in, worker.name);
	if (sqlite3_exec(db, sqlstr, NULL, NULL, &errmsg) != 0)
	{
        bzero(buf, N);
        strcpy(buf, "sign_in failed!");
        send(acceptfd, buf, N, 0);
		printf("error : %s\n", sqlite3_errmsg(db));
	}
	else 
    {
        bzero(buf, N);
        strcpy(buf, "sign_in ok!");
        send(acceptfd, buf, N, 0);
		printf("%s\n",buf);
    }
    return 0;
}

void do_quit(void)
{
    //pthread_exit("pthread_exit\n");
    //printf("exit succeed!\n");
    return;
}

/*功能选择函数
 *函数功能：接收客户端发来的操作请求，根据不同的操作请求调用相应的函数
 *函数参数：无
 *函数返回值：0
 * */
int function_select(void)
{
    while(1)
    {
        if(recv(acceptfd, op, N, 0) <= 0)
	    {
		    perror("fail to receive op\n");
    		exit(-1);
	    }
        //root用户的功能
        if(strncmp(op,"root_insert",11) == 0)
        {
            printf("server:excute root_insert function\n");
            do_insert(db);
        }
        if(strncmp(op,"root_delete",11) == 0)
        {
            printf("server:excute root_delete function\n");
            do_delete(db);
        }
        if(strncmp(op,"root_update",11) == 0)
        {
            printf("server:excute root_update function\n");
            do_root_update(db);
        }
        if(strncmp(op,"root_find",9) == 0)
        {
            printf("server:excute root_find function\n");
            do_root_find(db);
        }
        if(strncmp(op,"root_history",12) == 0)
        {
            printf("server:excute history function\n");
            do_root_history();
        }
        //普通用户的功能
        if(strncmp(op,"user_update",11) == 0)
        {
            printf("server:excute user_update function\n");
            do_user_update(db);
        }
        if(strncmp(op,"user_find",9) == 0)
        {
            printf("server:excute user_find function\n");
            do_user_find(db);
        }
        if(strncmp(op,"user_sign_in",12) == 0)
        {
            printf("server:excute user_sign_in function\n");
            do_user_sign_in(db);
        }
        if(strncmp(op,"quit",4) == 0)
        {
            printf("server:excute quit function\n");
            do_quit();
            printf("quit ok!\n");
            break;
        }
        do_log();
    }
    return 0;
}

/*线程回调函数
 *函数功能：接收来自客户端的登录信息结构体：login
 *          然后调用登录信息检查函数login_check()对登录信息进行检查
 *          检查无误后调用function_select()函数进行相应功能的选择
 *
 * */
void *pthread_fun(void *arg)
{
	if (sqlite3_open("empoloyee.db", &db) < 0)
	{
		printf("fail to sqlite3_open : %s\n", sqlite3_errmsg(db));
		exit(-1);
	}
    if(recv(acceptfd, (void *)&login, len_login, 0) <= 0)
	{
		perror("fail to receive\n");
		exit(-1);
	}
    printf("server:login_name:%s login_password:%s\n",login.name,login.password);

    login_check();

    function_select();

}

/*服务器端主程序
 * 函数功能：实现网络链接的初始化，客户端每发送一个连接请求
 *           就创建一个线程，由此实现并发；
 * */
int main(int argc, const char *argv[])
{
	struct sockaddr_in serveraddr, clientaddr;
	int len = sizeof(serveraddr);
    int len_worker = sizeof(worker);
    pthread_t tid;//线程标识符

	sockfd = socket(AF_INET, SOCK_STREAM, 0);
	if (sockfd < 0)
	{
		perror("fail to socket");
		exit(-1);
	}

	bzero(&serveraddr,len);//将地址结构中的成员置零

	serveraddr.sin_family = AF_INET;//ipv4
	serveraddr.sin_port = htons(atoi(argv[2]));
	serveraddr.sin_addr.s_addr = inet_addr(argv[1]);

	if (bind(sockfd, (struct sockaddr*)&serveraddr, len) < 0)
	{
		perror("fail to bind");
		exit(-1);
	}

	if (listen(sockfd, 5) < 0)
	{
		perror("fail to listen");
		exit(-1);
	}

	while(1)
    {
	    acceptfd = accept(sockfd, (struct sockaddr*)&clientaddr,&len);
	    if (acceptfd < 0)
	    {
		    perror("fail to accept");
		    exit(-1);
	    }

	    printf("addr = %s, port = %d \n", inet_ntoa(clientaddr.sin_addr),ntohs(clientaddr.sin_port));
        
        if(pthread_create(&tid,NULL,pthread_fun,NULL) !=0)
        {
            perror("fail to pthread_create!\n");
            exit(-1);
        }
    }
	return 0;
}


员工管理系统源码/client.c
#include "head.h"
struct worker_info_t worker;//员工信息结构体
struct login_info_t login;//登录信息结构体
int len_worker = sizeof(worker);
int len_login = sizeof(login);
struct select select_t;
int sockfd;
char buf[N] = {};
char func[N];
/*******************************************************
 *登录函数
 *函数功能：从客户端获取用户名和密码，将其封装进login结构体
 发送到服务器进行验证，并判断用户权限：
 root用户：login.flags = 1;
 普通用户：login.flags = 0;
 登录失败：login.flags = -1;
 *参数：sockfd；套接字描述符
 *返回值：0 登录成功  
 *        -1 登录失败； 
 * *****************************************************/
int do_login(int sockfd)
{
    char name_input[N] = {};
    char password_input[N] = {};
    printf("please input your name:");
    scanf("%s",name_input);
    printf("please input your password:");
    scanf("%s",password_input);
    strcpy(login.name,name_input);
    strcpy(login.password,password_input);
    if(send(sockfd,(void *)&login, len_login, 0) <= 0)
    {
        perror("fail to send login information\n");
        return -1;
    }
    if(recv(sockfd,(void *)&login, len_login, 0) <= 0)
    {
        perror("fail to recv login_flags\n");
        return -1;
    }
    return 0;
}
int do_insert(void)
{
    strcpy(func, "root_insert");
    send(sockfd,func, N, 0);

    printf("input name : ");
    scanf("%s", worker.name);
    printf("input sex : ");
    scanf("%s", worker.sex);
    printf("input age : ");
    scanf("%d", &(worker.age));
    printf("input id : ");
    scanf("%d", &(worker.id));
    printf("input section : ");
    scanf("%s", worker.section);
    printf("input salary : ");
    scanf("%d", &(worker.salary));
    printf("input sign_in : ");
    scanf("%d", &(worker.sign_in));
    printf("input performance : ");
    scanf("%d", &(worker.performance));
    if(send(sockfd, (void *)&worker, len_worker, 0) <= 0)
    {
        perror("fail to send worker message!\n");
        return -1;
    }

    printf("input password : ");
    scanf("%s", login.password);
    strcpy(login.name, worker.name);
    if(send(sockfd, (void *)&login, len_login, 0) <= 0)
    {
        perror("fail to send login message!\n");
        return -1;
    }

    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv insert result!\n");
        return -1;
    }
    printf("%s\n",buf);
    return 0;
}
int do_delete(void)
{
    strcpy(func, "root_delete");
    send(sockfd,func, N, 0);

    printf("input the name you want delete: ");
    scanf("%s", buf);
    if(send(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to send delete message!\n");
        return -1;
    }
    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv delete result!\n");
        return -1;
    }
    printf("%s\n",buf);

    return 0;
}

int do_root_update(void)
{
    strcpy(func, "root_update");
    send(sockfd,func, N, 0); 

	printf("**************** please select *****************\n");
	printf(" 1.update id. 2.update name. 3.update age.\n");
	printf("************************************************\n");
	printf("please input num ；");
	scanf("%d",&select_t.num);
	switch (select_t.num)
	{
	case 1 :
		printf("input new_id and old_name : ");
		scanf ("%d %s", &select_t.id, select_t.o_name);
		break;
	case 2 :
		printf("input new_name and old_name : ");
		scanf ("%s %s", select_t.n_name, select_t.o_name);
		break;			
	case 3 :
		printf("input new_age and old_name : ");
		scanf ("%d %s", &select_t.age, select_t.o_name);
		break;
	}
	if(send(sockfd,(void *)&select_t, sizeof(select_t), 0) <= 0)
    {
        perror("fail to send num information\n");
        return -1;
    }
    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv root_update result!\n");
        return -1;
    }
    printf("%s\n",buf);

    return 0;
}
int do_root_find(void)
{
    int num;
    strcpy(func, "root_find");
    send(sockfd,func, N, 0); 
#if 1
    bzero(buf, N);
    printf("input the name you want to find:\n");
    scanf("%s",buf);
    send(sockfd,buf, N, 0); 
#endif
    if(recv(sockfd, (void *)&worker, len_worker, 0) <= 0)
    {
        perror("fail to recv root_find information!\n");
        return -1;
    }
    printf("name sex age id section salary signin performance\n");
    printf("%-5s",worker.name);
    printf("%-5s",worker.sex);
    printf("%-5d",worker.age);
    printf("%-5d",worker.id);
    printf("%-5s",worker.section);
    printf("%-5d",worker.salary);
    printf("%-5d",worker.sign_in);
    printf("%-5d\n",worker.performance);

    return 0;
}
/*
 * 查看日志
 */
int do_root_history(void)
{
    strcpy(func, "root_history");
    send(sockfd,func, N, 0); 
	
    char buf[32] = {0};
	int ret;
	strcpy(func, "history");
	send(sockfd,func, N, 0);
	if(send(sockfd, (void *)&worker, len_worker, 0) < 0)
	{
		perror("fail to send show_history worker");
		return -1;
	}

	while((ret = recv(sockfd, buf, 31, 0)) > 0)
	{
		if(strncmp(buf,"end",3) == 0)
		{
			break;
		}
		printf("%s",buf);
	}
    return 0;
}

int do_user_update(void)
{
    strcpy(func, "user_update");
    send(sockfd,func, N, 0); 

	strcpy(select_t.o_name, login.name);
	printf("**************** please select *****************\n");
	printf(" 1.update id. 2.update password. 3.update age.\n");
	printf("************************************************\n");
	printf("please input num ；");
	scanf("%d",&select_t.num);
	switch (select_t.num)
	{
	case 1 :
		printf("input new_id : ");
		scanf ("%d", &select_t.id);
		break;
	case 2 :
		printf("input new_password : ");
		scanf ("%s", select_t.n_name);
		break;			
	case 3 :
		printf("input new_age : ");
		scanf ("%d", &select_t.age);
		break;
	}
	if(send(sockfd,(void *)&select_t, sizeof(select_t), 0) <= 0)
    {
        perror("fail to send num information\n");
        return -1;
    }
    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv user_update result!\n");
        return -1;
    }
    printf("%s\n",buf);
    return 0;
}

int do_user_find(void)
{
    strcpy(func, "user_find");
    send(sockfd,func, N, 0); 

    if(recv(sockfd, (void *)&worker, len_worker, 0) <= 0)
    {
        perror("fail to recv user_find result!\n");
        return -1;
    }
    printf("name sex age id section salary signin performance\n");
    printf("%-5s",worker.name);
    printf("%-5s",worker.sex);
    printf("%-5d",worker.age);
    printf("%-5d",worker.id);
    printf("%-5s",worker.section);
    printf("%-5d",worker.salary);
    printf("%-5d",worker.sign_in);
    printf("%-5d\n",worker.performance);

    return 0;
}

int do_user_sign_in(void)
{
    strcpy(func, "user_sign_in");
    send(sockfd,func, N, 0); 
    
	time_t tm;
	struct tm *mytm;
	time(&tm);  //获取秒数
	mytm = localtime(&tm);//将描述转化为年月日十分秒
	
    /**将早上8:30-8:59规定为签到时间**/
    if((mytm->tm_hour == 8) && ((mytm->tm_min >=30) && (mytm->tm_min <= 59)))
    {
        worker.sign_in = 1;
    }
    else
    {
        worker.sign_in = 0;
    }
	if(send(sockfd,(void *)&worker, len_worker, 0) <= 0)
    {
        perror("fail to send signin information\n");
        return -1;
    }
    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv signin result!\n");
        return -1;
    }
    printf("%s\n",buf);
    
    return 0;
}

int do_quit(void)
{
    strcpy(func, "quit");
    send(sockfd,func, N, 0); 
#if 0
    bzero(buf, N);
    if(recv(sockfd, buf, N, 0) <= 0)
    {
        perror("fail to recv quit result!\n");
        return -1;
    }
    printf("%s\n",buf);
#endif
    exit(0);
}

/******************************************
 *用户界面选择函数
 *函数功能：根据不同的权限选择不同的用户界面
 *参数：无
 *返回值：无
 * *****************************************/
int choose_interface(void)
{
    char op;
    
    while(1)
    {
        if(login.flags == 1)
        {
            printf("login success! you are a root user!\n");
            printf("**************************************\n");
            printf("*       1 ---> insert a user         *\n");
            printf("*       2 ---> delete a user         *\n");
            printf("*       3 ---> update a user         *\n");
            printf("*       4 ---> find  a  user         *\n");
            printf("*       5 ---> history               *\n");
            printf("*       6 ---> quit                  *\n");
            printf("**************************************\n");
            printf("choose your operation:\n");
            getchar();
            op = getchar();
            switch(op)
            {
                case'1':
                    printf("client: excute root_insert function!\n");
                    do_insert();
                    break;
                case'2':
                    printf("client: excute root_delete function!\n");
                    do_delete();
                    break;
                case'3':
                    printf("client: excute root_update function!\n");
                    do_root_update();
                    break;
                case'4':
                    printf("client: excute root_find function!\n");
                    do_root_find();
                    break;
                case'5':
                    printf("client: excute root_history function!\n");
                    do_root_history();
                    break;
                case'6':
                    printf("client: excute root_quit function!\n");
                    do_quit();
                    break;
                default:
                    return -1;
            }
        }
        else if(login.flags ==0)
        {
            printf("login success! you are a normal user!\n");
            printf("**************************************\n");
            printf("*       1 ---> find my information   *\n");
            printf("*       2 ---> update my information *\n");
            printf("*       3 ---> sign in               *\n");
            printf("*       4 ---> quit                  *\n");
            printf("**************************************\n");
            printf("choose your operation:\n");
            getchar();
            op = getchar();
            switch(op)
            {
                case'1':
                    printf("client: excute user_find function!\n");
                    do_user_find();
                    break;
                case'2':
                    printf("client: excute user_update function!\n");
                    do_user_update();
                    break;
                case'3':
                    printf("client: excute user_sign_in function!\n");
                    do_user_sign_in();
                    break;
                case'4':
                    printf("client: excute user_quit function!\n");
                    do_quit();
                    break;
                default:
                    return -1;
            }
        }
        else if(login.flags == -1)
        {
            printf("this user doesn't exist,please input again!\n");
            return -1;
        }
        else
        {
            printf("invalid login flags\n");
            return -1;
        }
    }
    return 0;
}
/********************************
 *主函数
 *函数功能：初始化客户端的网络连接
 * *******************************/
int main(int argc, const char *argv[])
{
    struct sockaddr_in serveraddr;
    int len = sizeof(serveraddr);
    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0)
    {
        perror("fail to socket");
        exit(1);
    }
    bzero(&serveraddr,len);
    serveraddr.sin_family = AF_INET;
    serveraddr.sin_port = htons(atoi(argv[2]));
    serveraddr.sin_addr.s_addr = inet_addr(argv[1]);
    //向服务器发送链接请求
    if (connect(sockfd,(struct sockaddr*)&serveraddr, len) < 0)	
    {
        perror("fail to connect");
        exit(1);
    }
    //用户登录
login:
    if(do_login(sockfd) < 0)
    {
        printf("fail to login\n");
        goto login;
    }
    //选择用户界面
    choose_interface();
    while(1);
    return 0;
}
